Support for cloud microphysics hail species

MicroTed opened the pull request:
**Description**

Adds support for hail hydrometeor category (nwat = 7) for 7-class cloud microphysics. Specifically for the NSSL microphysics scheme, but should work for any microphysics that has separate graupel and hail/frozen drops categories.

Related PR for CCPP physics: https://github.com/NCAR/ccpp-physics/pull/761
Related PR for Single Column Model: https://github.com/NCAR/ccpp-scm/pull/277
Related PR of fv3atm: https://github.com/NOAA-EMC/fv3atm/pull/472

**How Has This Been Tested?**

Tested to work with fv3atm develop branch on Macos (gcc-10) and Unix (NOAA/Jet, Intel compilers)
Cases run with "grad student test" case (2019061518) and RRFS domain tests
Has *not* yet been ported to dev/gfdl branch, but fv3atm should still compile and run with other suites that have nwat <= 6.

**Checklist:**

- [ x] My code follows the style guidelines of this project
- [ x] I have performed a self-review of my own code
- [ x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ x] My changes generate no new warnings
- [ x] Any dependent changes have been merged and published in downstream modules

Changes follow or are similar to previous work by Tim Supinie @tsupinie, who deserves much of the credit here. Also contributions by @LarissaReames-NOAA

lharris4 commented on model/fv_dynamics.F90:
```diff
@@ -849,7 +862,44 @@ subroutine fv_dynamics(npx, npy, npz, nq_tot,  ng, bdt, consv_te, fill,
   endif
 
-  if( nwat==6 ) then
+  if( nwat==7 ) then
```
This is the right way to add a new nwat to the code. I am hoping that eventually there can be a more generic way to do this though so that we don't need separate hard codings for different numbers of ice categories.

lharris4 replied:
That would work. There should be an easy way for the physics to query the dynamics for the nwat value.

MicroTed replied:
@bensonr Right, nwat is being passed to model%init as intent(in). It looks like it is used before that, so it might take some changes to be able to set it in model%init. Something to look at for a future update?

lharris4 reviewed (approved):
This is a good addition that provides essential functionality (more schemes now have separate hail and graupel categories, something that may be necessary in future global storm resolving models) and correctly follows existing code standards. I approve this merge if it passes the appropriate regression tests.

bensonr commented on tools/external_ic.F90:
```diff
@@ -937,8 +955,15 @@ subroutine read_gfs_ic()
           do nt = 1, ntracers
-             q(:,:,:,nt) = -999.99
+
+             IF ( .not. ( nt == Atm%flagstruct%nwat .and. Atm%flagstruct%nwat == 7 ) ) THEN
```
This is a necessary trigger to ensure proper initialization of tracer quantities when the experiment defines more tracers than are available in the IC files.  The field_table has attribute modifiers that allow one to specify the initial profile for a tracer.  By having the temporary 'q' define all non-existent IC tracers to zero, you are replacing any initial field_table profiles defined at lines 639-648 for Atm%q and Atm%qdiag.  

If you have non-IC file tracers you want to be zero, the proper method is to use the attributes correctly within the field_table.  The -999.99 value works with the logic at line 3087 to the non-existent IC tracers and also ensures there is no potential round-off errors brought in by the remapping and 'fill' process.  If you still have questions, we can discuss further.

MicroTed replied:
The negative value initialization seems to have been added some time in the past year or so? It is only the hail mass that has a problem here, since that is not in the inputs I have used. Is there a reason not to simply set everything to zero, or was this meant to trap some particular issue?

MicroTed replied:
Ah, OK, thanks for the explanation. I will revert that and test with changing the field_table.

bensonr replied:
As stated above, not all tracer initial values should be set to zero.  The q(:)=0. logic you added forces all tracers to be zero if they don't have an entry in the IC files.  If you want the hailwat tracer to be defined as zero, a properly configured entry for the hail tracer in the field_table will ensure it's initial value in Atm%q will be so and the value of q(halwat) = -999.99 and the if-test at line 3087 is left in place.

MicroTed commented on model/fv_sg.F90:
```diff
@@ -1878,6 +1904,364 @@ subroutine neg_adj3(is, ie, js, je, ng, kbot, hydrostatic, peln, delz, pt, dp,
 
  end subroutine neg_adj3
 
+! TAS: XXX check to make sure this doesn't need any modifications vs neg_adj2 or neg_adj3
```
Yes, good catch. (@tsupinie - thoughts?) The microphysics scheme here should not output negative hail mass values (below-threshold mass values are moved to water vapor), we could call neg_adj3 instead? Should we assume the same for other new microphysics schemes that have hail (nwat=7)?

lharris4 commented:
Hi, Ted. Another thing you can do is to check if the 'hailwat' tracer is
defined. If the tracer is not defined then get_tracer_index('hailwat') will
return a negative value.

Lucas

On Mon, Jan 31, 2022 at 11:49 AM Ted Mansell ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In model/fv_regional_bc.F90
> <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/171#discussion_r795865027>
> :
>
> > @@ -3808,7 +3811,51 @@ subroutine remap_scalar_nggps_regional_bc(Atm                         &
>           enddo
>        enddo
>     endif
> -  endif ! data source /= FV3GFS GAUSSIAN NEMSIO/NETCDF and GRIB2 FILE
> +   if ( Atm%flagstruct%nwat .eq. 7 ) then
> +      do k=1,npz
> +         do i=is,ie
> +            qn1(i,k) = BC_side%q_BC(i,j,k,liq_wat)
> +            BC_side%q_BC(i,j,k,rainwat) = 0.
> +            BC_side%q_BC(i,j,k,snowwat) = 0.
> +            BC_side%q_BC(i,j,k,graupel) = 0.
> +            BC_side%q_BC(i,j,k,hailwat) = 0.
> +            if ( BC_side%pt_BC(i,j,k) > 273.16 ) then       ! > 0C all liq_wat
>
> Looking again at this, a separate nwat=7 section seems unnecessary. There
> can be just one section with
>
> if ( Atm%flagstruct%nwat .eq. 6 .or. Atm%flagstruct%nwat .eq. 7 ) then
> if ( Atm%flagstruct%nwat .eq. 7 ) BC_side%q_BC(is:ie,j,1:npz,hailwat) = 0.
> do k=1,npz
> ...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/pull/171#discussion_r795865027>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AMUQRVD4D4YHCZRWC7V2DQ3UY24RZANCNFSM5NB6624Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

MicroTed commented on tools/external_ic.F90:
```diff
@@ -800,6 +804,13 @@ subroutine get_nggps_ic (Atm, fv_domain)
+                           Atm%q(i,j,k,rainwat) + &
+                           Atm%q(i,j,k,snowwat) + &
+                           Atm%q(i,j,k,graupel) + &
+                           Atm%q(i,j,k,hailwat)))
```
Code removed!

MicroTed commented:
> Another thing you can do is to check if the 'hailwat' tracer is defined. If the tracer is not defined then get_tracer_index('hailwat') will return a negative value

Good point, Lucas. Although currently the assumption is that hail would only be in a 7-class scheme, future ones may not conform to that, so switching toward checking for categories generally may make it easier to incorporate options like P3 in the future. I have experimented with using a combined ice+snow category, as well. This particular section seems to be set up for a special case (expecting only nwat=6), so the mods are just to be sure all bases are covered.

lharris4 reviewed (approved):
The new updates are good. Thanks.

laurenchilutti commented:
@junwang-noaa This PR has been fully reviewed and is ready to be added to the commit queue and merge.  Please let me know when you would like me to merge.

grantfirl commented:
@MicroTed Since #176 was merged into dev/emc, you will need to merge in the latest dev/emc to your branch please.

MicroTed commented:
OK, I have updated to dev/emc, but I'm having trouble getting the whole thing to compile (e.g., module_wrt_grid_comp.F90) (also updated fv3atm to current develop)

laurenchilutti commented:
@junwang-noaa I believe this is on the commit queue for today.  Can I merge this now?

junwang-noaa commented:
@laurenchilutti Would you please merge this PR? Thanks
